### PR TITLE
[model] fix: fix vocab_weights device and shard handling for qwen3_vl.py

### PR DESCRIPTION
**Pull request by ocp8**
### What does this PR do?

Now qwen3_5 has a shard handling for vocab_weights(self.lm_head.weight) when it is a sharded tensor. Copy the same for qwen3_vl, and also solve the device problem when it is used with torch.compile which makes it as a cpu tensor by default

Env:
- Verl 0.8.0
- Torch 2.11.0
- Cu130

Relevant issue: https://github.com/verl-project/verl/issues/4756 

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always
- pass

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: no need to add more tests
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/models/transformers/qwen3_vl.py [hunk 1]
```diff
@@ -364,11 +364,15 @@ def forward_with_torch_backend(
         rolled_labels = torch.roll(input_ids, shifts=-1, dims=-1)
     else:
         raise RuntimeError("To use forward_with_torch_backend, either labels or input_ids must be provided.")
+        
+    vocab_weights = self.lm_head.weight
+    if isinstance(vocab_weights, DTensor):
+        vocab_weights = vocab_weights.full_tensor().to(hidden_states.device)
```
[the remark this answers is not in the document]
Comment (ocp8) [reply verl/models/transformers/qwen3_vl.py:371]: Added missing import command

**Comment by Luosuu** [comment]
Hi @ocp8 

In what case we will run into issues without this fix?

**Comment by ocp8** [comment]
@Luosuu  sorry I missed this msg before.

It happens when I am running multimodal GRPO on Qwen3 and Qwen3.5 (multiple images in one prompt, but I guess it doesn't matter). 

**Env**:
Verl 0.8.0
Torch 2.11.0
Cu130

**Command**: 
python3 -m verl.trainer.main_ppo 
algorithm.adv_estimator=grpo

I remember the official example code can also realize the same issue on my machine.

**Comment by ocp8** [comment]
By the way I am not using Megatron, but FSDP. And I encountered this issue when I am only using one card (all parallelism = 1), maybe that is the cause

**Comment by Luosuu** [comment]
@ocp8 thanks for the clarification. what confused me is that this PR tries to fix a distributed tensor issue but you say this is only encountered with 1 GPU. Recently the distributed Qwen3.5 FSDP runs are also healthy so I am not sure what his actually fixes

**Comment by ocp8** [comment]
> @ocp8 thanks for the clarification. what confused me is that this PR tries to fix a distributed tensor issue but you say this is only encountered with 1 GPU. Recently the distributed Qwen3.5 FSDP runs are also healthy so I am not sure what his actually fixes

@Luosuu yes that's the reason then. I don't know where the root cause is (I guess in fsdp) but it does happen in the single card case. I guess there are some people like me who only has one card but still wants to run RL so maybe it is still worthy lol. 

let me know if anything else I can help.

**Comment by Luosuu** [comment]
@ocp8 would you like to figure out why and what happened in the single card case and why this can fix?

**Comment by ocp8** [comment]
> @ocp8 would you like to figure out why and what happened in the single card case and why this can fix?

@Luosuu I tried to deep dive:
- I had a look at the trace. It happens because hidden_states is a Tensor while vocab_weights is a DTensor. 
- It happens because self.lm_head.weight is a DTensor (passed at https://github.com/verl-project/verl/blob/c165b3da1f660cfff83f9afe23654346dadacdd7/verl/models/transformers/qwen3_vl.py#L371)
- self.lm_head.weight is a DTensor because in FSDP2, the weights are saved as DTensor. 
- While during a normal forward of self.lm_head(nn.Linear), FSDP2 adds a pre-forward hook to unshard the weights and do the calculation. However, if we directly take it as the function input above, it is still a DTensor and cause the error.
- Therefore, it is not limited to one card only (just because I had only one card before). I tried to use 2 cards on single machine and I can still reproduce this error on qwen3vl.
- It works for qwen3.5 because qwen3.5 already has this sharding fix. Qwen3vl doesn't. (https://github.com/verl-project/verl/blob/c165b3da1f660cfff83f9afe23654346dadacdd7/verl/models/transformers/qwen3_5.py#L283)

**Review by Luosuu — DISMISSED** [review]

**Comment by Luosuu** [comment]
@ocp8 thanks very much for the investigation. I have triggered CI for you and please fix the pre-commit 
 feel free to @ me if I forget to re-trigger CI to merge this PR

**Comment by ocp8** [comment]
> @ocp8 thanks very much for the investigation. I have triggered CI for you and please fix the pre-commit feel free to @ me if I forget to re-trigger CI to merge this PR

@Luosuu I fixed the ruff format error in the pre-commit. Pls kindly re-trigger. Thanks.

**Comment by ocp8** [comment]
@Luosuu the error in previous pre-commit check has been solved, and it looks like the rest 7 failed checks are not relevant to this pr. Pls let me know if anything else I could help :)

**Comment by Luosuu** [comment]
@ocp8 thanks. let me rerun gpu unit tests

**Review by Luosuu — APPROVED** [review]
